src/cloudflare/internal/workflows-api.ts

```diff
@@ -1,12 +1,173 @@
+  // `__brand` is needed for engine to validate if the user returned a NonRetryableError
+  // This provides a better DX because they can just extend NonRetryableError for their own Errors
+  // and override name
+  // Private fields are not serialized over RPC
   public readonly __brand: string = 'NonRetryableError';
 
   public constructor(message: string, name = 'NonRetryableError') {
     super(message);
     this.name = name;
   }
 }
+
+type InstanceStatus = {
+  status:
+    | 'queued'
+    | 'running'
+    | 'paused'
+    | 'errored'
+    | 'terminated'
+    | 'complete'
+    | 'unknown';
+  error?: string;
+  output?: object;
+};
+
+interface Fetcher {
+  fetch: typeof fetch;
+}
+
+async function callFetcher<T>(
+  fetcher: Fetcher,
+  path: string,
+  body: object
+): Promise<T> {
+  const response = await fetcher.fetch(`http://workflow-binding.local${path}`, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      'X-Version': '1',
+    },
+    body: JSON.stringify(body),
+  });
+
+  // TODO: Throw if result is not 200 or ok is false
+
+  return response.json() as Promise<T>;
+}
+
+export class Instance {
+  private readonly fetcher: Fetcher;
+  private readonly instanceId: string;
+
+  public constructor(instanceId: string, fetcher: Fetcher) {
+    this.instanceId = instanceId;
+    this.fetcher = fetcher;
+  }
+
+  public async pause(): Promise<void> {
+    try {
+      await callFetcher(this.fetcher, '/pause-instance', {
+        instanceId: this.instanceId,
+      });
+    } catch (_error) {
+      throw new WorkflowError(
+        'Failed to pause Workflow instance ' + this.instanceId
+      );
+    }
+  }
+  public async resume(): Promise<void> {
+    try {
+      await callFetcher(this.fetcher, '/resume-instance', {
+        instanceId: this.instanceId,
+      });
+    } catch (_error) {
+      throw new WorkflowError(
+        'Failed to resume Workflow instance ' + this.instanceId
+      );
+    }
+  }
+
+  public async abort(): Promise<void> {
+    try {
+      await callFetcher(this.fetcher, '/abort-instance', {
+        instanceId: this.instanceId,
+      });
+    } catch (_error) {
+      throw new WorkflowError(
+        'Failed to abort Workflow instance ' + this.instanceId
+      );
+    }
+  }
+
+  public async restart(): Promise<void> {
+    try {
+      await callFetcher(this.fetcher, '/restart-instance', {
+        instanceId: this.instanceId,
+      });
+    } catch (_error) {
+      throw new WorkflowError(
+        'Failed to restart Workflow instance ' + this.instanceId
+      );
+    }
+  }
+
+  public async status(): Promise<InstanceStatus> {
+    try {
+      const result = await callFetcher<InstanceStatus>(
+        this.fetcher,
+        '/instance-status',
+        {
+          instanceId: this.instanceId,
+        }
+      );
+      return result;
+    } catch (_error) {
+      throw new WorkflowError(
+        'Failed to get status for Workflow instance ' + this.instanceId
```

Comment:
als _all_ of these look the same. Maybe `callFetcher` can accept error message and throw `WorkflowError` instead?

Reply:
Yeah, I should probably just do that. Good thinking.